### ISSUE TITLE
License addition criterion 3 disregards the decentralized nature of git and open source

### ISSUE THREAD

**Issue by simonrepp** [github/choosealicense.com#739]
The [Adding a license](https://github.com/github/choosealicense.com/blob/gh-pages/CONTRIBUTING.md#adding-a-license) page states criteria for the addition of new licenses to https://choosealicense.com, among these it reads:

> 3. A GitHub code search must reveal at least 1,000 public repositories using the license.

At this point I'd like to quote the tagline of this repository:

> choosealicense.com - A site to provide non-judgmental guidance on choosing a license for your open source project

If the site is non-judgemental about licenses, I think it should also not judge their relevance based on one single platform in a much larger ecosystem, the site is after all called "choosealicense.com", not "chooseagithublicense.com".

Furthermore I just issued a [search for "open source licenses" on google](https://www.google.com/search?hl=en&q=open%20source%20licenses) and choosealicense.com came up second in my results - which demonstrates the popularity and influence this site has gained (for which I applaud everyone who has contributed to make it so), but I firmly believe also implies a responsibility to offer a fair representation of licenses regardless of their popularity on github.com in particular in order to allow new, innovative licenses to have a voice and gain popularity through choosealicense.com as well.

Personally I believe 1,000 as a threshold to be unnecessarily high and more damaging to innovation than protective of "spam", but regardless of that **I hereby propose to make this criterion (whatever the number) at least respect and value the decentralized nature of git and open source and therefore count the license usage among all the diverse platforms and communities out there**.

Proposing a new license, given the tough criteria, is in itself quite an undertaking, I therefore see no relevant possibility for this loosening of criteria to be exploited in any way. A malign attempt to conjure up a fake count of usage through mock platforms created for only this purpose would easily be spotted in qualitative review and its rejection surely would be understood and supported by the community at large anyway, even if the criteria technically allow some leeway.

I'd like to thank everyone for their work on choosealicense.com and would ask for due consideration of this issue. I can gladly contribute a concrete proposal and PR for this if needed to make this happen from a practical point of view.

**Comment by waldyrious** [comment]
@simonrepp I fully agree that it would make sense to include more code hosts in the criterion for measuring a license's adoption. But it's not just a matter of rewording that criterion — we need to provide an easy way to validate it.

GitHub ends up being a reasonable proxy for the adoption metric, since it's the most popular host, and therefore one would assume that, by the law of large numbers, its distribution of licenses would approximate the true distribution when considering all code hosts out there; that, coupled with the fact that it's easy to search for [repositories filtered by license](https://github.com/search?q=license%3Aisc&type=Repositories), makes it an imperfect but pratical heuristic for this effect.

What would you suggest as an alternative technique to measure this criterion in a broader universe of repositories? Perhaps we could also incorporate a similar filtered search on GitLab and Bitbucket, and manually sum the results? Or do you know of an aggregation platform that allows searching across multiple code hosts while filtering by license?

**Comment by simonrepp** [comment]
@waldyrious Thanks for sharing your thoughts! 

My main idea here is to not overengineer/-specify the requested proof of relevance for applications - 1,000 repositories is a huge number anyway and whoever attains that kind of adoption will probably be able to compile a reasonable listing (which as you propose sums up to a total) from a few sources/platforms, and most likely that will be in the form of a few search urls on different sites, although I don't think this needs to be explicitly specified. I believe it's reasonable to assume that someone who wrote a license that was picked up by 1,000 different repositories and reviewed e.g. by OSI to provide that proof and listing in a sensible form that is easily reviewed. If an application for addition comes in an unacceptable/unreviewable form this can be given as feedback and a more sensible application requested. (but I think this is rather hypothetical) Also I would strongly advise to not specify a fixed list of platforms, as I'd rather not open another issue in 2 years when it becomes obsolete due to rapid changes in the open source landscape. :)

I disagree that github is a reasonable proxy by the way - I think it's fair to state (and really I'm doing this without any moral judgement here) that there is a considerable and growing amount of communities and people that for various motives avoid hosting their code on github (or other large platforms for that matter). If there is a correlation with a favor in licenses in these communities it essentially introduces a bias, the avoidance of which is the entire point of the issue I created.

So to sum it up from my side - I don't see a need for contriving a specific format or technical metric to prove this quantitative threshold (1,000) set by the guidelines of choosealicense.com. I think this is mostly a matter of just agreeing to revise the formulation to make it platform-independent. The search url approach which is currently mandatory could be turned into a technical suggestion instead, aiding people to prepare their proposal (search urls for other platforms could be provided too), but in any case this should be free-form and trust be put into people to come up with something reasonable on their own.

**Comment by waldyrious** [comment]
> there is a considerable and growing amount of communities and people that for various motives avoid hosting their code on github (or other large platforms for that matter). If there is a correlation with a favor in licenses in these communities it essentially introduces a bias

I was concerned about that as well as I wrote my comment above. However, I think you would agree that including the 3 or 4 major platforms would be sufficient to include most of such communities, at least to the degree that they may affect the scale of adoption we're talking about here. If a hyper-specialized community prefers obscure platform X and uncommon license Y, it doesn't seem likely that they would tilt the scales significantly even if they were included in the count. (And in any case, that would be a data point which I'm sure the maintainers would take into consideration if presented with it, in the unlikely cases that it _would_ make a difference.)

> My main idea here is to not overengineer/-specify the requested proof of relevance
> ...
> I don't see a need for contriving a specific format or technical metric to prove this quantitative threshold

Quite the contrary: by making it specific we make it clear what data collection effort is expected, are even able to provide demo search URLs that both contributors and maintainers can quickly adapt to assess a proposed license's adherence to this criterion. I see the goal of making it specific as a way to make thing easier, not harder.

And as you say, the panorama will certainly change after a while, at which point we can make adjustments just like the ones we're discussing here. I am sure you can appreciate that no specification will be 100% future-proof, so we might as well make everyone's lives easier while the current conditions apply.

**Comment by mlinksva** [comment]
Choosealicense.com is really not the place "to allow new, innovative licenses to have a voice and gain popularity". 1000 repos is a very high bar for a random new OSS license, of which there are many, but if such a new license _really_ adds value, it can be demonstrated by such adoption, and will blow past 1000 repos.

I'm fine with considering evidence of adoption in repos on other public hosts, including non-git ones. I'll tweak the language in CONTRIBUTING to allow for that, while keeping the concrete GitHub example.

Aside, @waldyrious you wrote:

> Or do you know of an aggregation platform that allows searching across multiple code hosts while filtering by license?

I don't, but would enjoy knowing of one, or things in that direction. Sort of in that direction that I know of:

- [Wikidata query](https://w.wiki/SfM) counting uses of licenses for items with a repo and a license that has an SPDX ID (I'm not at all confident it isn't more complicated than that, eg 0BSD [results](https://w.wiki/SfT) are fishy _added: the [reason](https://www.wikidata.org/wiki/Talk:Q191307#SPDX_IDs)_)
- [Searchcode](https://searchcode.com/?q=%22permission+is+hereby+granted%22&lan=232) does aggregation, but seems to be doing some kind of substring search that doesn't work well for searching for unique fragments needed to identify likely instances of specific licenses
- [Software Heritage](https://www.softwareheritage.org/) is doing aggregation but as far as I know does not allow searching contents at this point

**Comment by simonrepp** [comment]
@waldyrious 

> I see the goal of making it specific as a way to make thing easier, not harder.

For specific url *suggestions* I fully agree, that will make it easy for the suggested platforms themselves, and easier to deduce method for those not listed, which is great! A *fixed set* of specified (and therefore exclusively allowed) platforms however, will make it easy for the listed ones and *impossible* for anything that is not considered by us now, meaning again someone like me will have to open an issue, it will need to be discussed again ... I'm not sure what the advantage would be of that to just leaving this open for other platforms.

No specification is 100% future-proof, I agree, but an exclusive list of platforms would not even be 100% present-proof, as it would only get as good as our small collective picture of what we think is out there that is relevant, and that might not even fairly represent some platforms and communities we don't know about that are out there *right now*. I do a good amount of research on these things and yet I can not claim I could fairly write that list right now (or ever really).

@mlinksva 

> Choosealicense.com is really not the place "to allow new, innovative licenses to have a voice and gain popularity". 

Ok noted. Personally I believe it should be though - if a platform grows powerful enough to significantly influence culture and decisionmaking for millions of people entering a field, it should take responsibility to allow that culture and decisionmaking to develop further as well, that's also why I don't think the 1,000 limit is a good idea either, but as I mentioned that's more of an aside.

I think an important step is made if at least the platform lock-in is gone, and I'm thankful that we seem to agree on that point, thank you for the support on that, to both of you. :+1:
